File: 20.有效的括号.go
package leetcode

/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */

// @lc code=start
// s[i] 可以取string中的byte出来
func isValid(s string) bool {
	m := make(map[byte]byte)
	m[')'] = '('
	m[']'] = '['
	m['}'] = '{'
	var stack []byte
	for i := 0; i < len(s); i++ {
		if len(stack) > 0 && m[s[i]] > 0 {
			if stack[len(stack)-1] != m[s[i]] {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}
	if len(stack) > 0 {
		return false
	}
	return true
}

// @lc code=end

File: 20.有效的括号.java
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Map<Character,Character> map = new HashMap<Character,Character>();
        map.put(')', '(');
        map.put(']', '[');
        map.put('}', '{');
        Deque<Character> deque = new LinkedList<Character>();
        for (int i = 0; i < s.length(); i++) {
            if(!map.containsKey(s.charAt(i))){
                deque.push(s.charAt(i));
            }else{
                if(map.get(s.charAt(i)) != deque.peek()){
                    return false;
                }
                deque.pop();
            }
        }
        if (deque.size()>0){
            return false;
        }
        return true;
    }
}
// @lc code=end

